src/library/equations_compiler/elim_match.cpp

```diff
@@ -502,17 +503,14 @@ struct elim_match_fn {
                 expr const & p = head(eqn.m_patterns);
                 if (is_local(p)) {
                     has_variable = true; return true;
-                } else if (is_constructor_app(p)) {
+                }
+                type_context ctx = mk_type_context(eqn.m_lctx);
+                bool is_val = is_value(ctx, p);
+                if ((is_constructor_app(p) && (value_is_contructor || !is_val)) ||
+                    (value_is_contructor && is_val)) {
                     has_constructor = true; return true;
-                } else {
-                    type_context ctx = mk_type_context(eqn.m_lctx);
-                    if (is_value(ctx, p)) {
-                        if (value_is_contructor) has_constructor = true;
-                        return true;
-                    } else {
-                        return false;
-                    }
                 }
+                return false;
             });
         return r && has_variable && has_constructor;
     }
```

Comment:
@kha Ok. So, I think we should remove `enum` inductive from `is_value`. That is, we remove the test `if (optional<name> I_name = is_constructor(e)) return is_nontrivial_enum(*I_name);` from `is_value`.  As you pointed out, the value transition was never being used for them anyway. 
After, we make this change, we can always try the value_transition before complete_transition, as you wanted.